### PR TITLE
1405 add bending energy loss

### PR DESCRIPTION
**Pull request by kate-sann5100**
Fixes #1408 .

### Description
Add bending energy loss useful for registration.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/losses/deform.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+from typing import Union
+
+import torch
+from torch.nn.modules.loss import _Loss
+
+from monai.utils import LossReduction
+
+
+def gradient_dx(fx: torch.Tensor) -> torch.Tensor:
```
Comment (wyli) [inline monai/losses/deform.py:9]: could unify dx/dy/dz and make it generic using the slicing objects:
```py
    slice_1 = slice(1, -1)
    slice_2_s = slice(2, None)
    slice_2_e = slice(None, -2)
    slice_all = slice(None)
    slicing_s = [slice_1, slice_1, slice_2_s]
    slicing_e = [slice_1, slice_1, slice_2_e]
    while len(slicing_s) < fx.ndim:
        slicing_s = [slice_all] + slicing_s
        slicing_e = [slice_all] + slicing_e
    return (fx[slicing_s] - fx[slicing_e]) / 2.0
```

File: monai/losses/deform.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        assert len(input.shape) == 5 and input.shape[1] == 3, (
+            f"expecting 5-d ddf input with 3 channels, " f"instead got input of shape {input.shape}"
+        )
+        assert (
+            input.shape[-1] > 4 and input.shape[-2] > 4 and input.shape[-3] > 4
+        ), f"all depth, height and width must > 4, got input of shape {input.shape}"
+
+        # first order gradient
+        # (batch, 3, d-2, h-2, w-2)
+        dfdx = gradient_dx(input)
```
Comment (wyli) [inline monai/losses/deform.py:87]: this needs to be extended to support 1d/2d/3d

**Review by wyli — COMMENTED** [review]
thanks, the style/format of this PR looks great, it's a really nice starting point. I put some comments about the actual logic. ideally we should have something general for different spatial dimensions, not just for spatial 3D

**Review by wyli — APPROVED** [review]
thanks, I forgot to mention that perhaps rename `gradient` to `spatial_gradient` to avoid confusion? 

also, do you want to put some acknowledgement in the docstring such as`adapted from DeepRegNet`? cc @YipengHu

**Comment by YipengHu** [comment]
> thanks, I forgot to mention that perhaps rename `gradient` to `spatial_gradient` to avoid confusion?
> 
> also, do you want to put some acknowledgement in the docstring such as`adapted from DeepRegNet`? cc @YipengHu

Sounds good to have a reference to DeepReg, using either the paper or github URL.
"spatial_gradient" is more specific and better.

File: monai/losses/deform.py (hunk 2 above)
Comment (YipengHu) [reply monai/losses/deform.py]: @kate-sann5100 in case needs a reference for 2D: https://weisslab.cs.ucl.ac.uk/YipengHu/mphy0041/-/blob/main/tutorials/registration/pt_utils.py

just curious what application is there using 1d spatial gradient :)
